### PR TITLE
Revolution Nodes

### PR DESCRIPTION
**Pull request by TrainWrack**
Please test them out. Attached is the test level.
[RevolutionTest.zip](https://github.com/user-attachments/files/16480797/RevolutionTest.zip)

### CODE REVIEW

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    local center2pos = center2:GetPosition()
+	local center2rot = center2:GetRotation()
+    local radius2 = LevelVars[dataName].Radius
+    local angle = newValue1 * math.pi / 180
+    local x = center2pos.x
+    local y = center2pos.y
+    local z = center2pos.z
+	local normalizedNewValue1 = NormalizeAngle(newValue1)
+    local normalizedEndAngle = NormalizeAngle(LevelVars[dataName].EndAngle)
+
```
Comment (Stranger1992) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua:75]: Remove Line here

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    LevelVars[dataName].StartAngle = startangle
+    LevelVars[dataName].EndAngle   = endangle
+    LevelVars[dataName].CCW        = isCCW
+    LevelVars[dataName].Loop       = isLoop
+    LevelVars[dataName].Smooth     = isSmooth
+    LevelVars[dataName].Rotate     = isRotate
+    LevelVars[dataName].CRotate    = isCRotate
+    LevelVars[dataName].Sfx        = SFX
+    LevelVars[dataName].OldValue   = startangle
+	LevelVars[dataName].StopAtEnd  = false
```
Comment (Stranger1992) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: Remove tabulation here (So everything is inline)

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    
+	LevelVars[dataName].NewValue = LevelVars[dataName].Loop
+    and (startangle + (LevelVars[dataName].CCW and -360 or 360))
```
Comment (Stranger1992) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: Ideally, declare the magic numbers:

`local rotationAngle = 360`

and then change line 39 to:

` and (startangle + (LevelVars[dataName].CCW and -rotationAngle  or rotationAngle ))`

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    return (angle % 360 + 360) % 360
+	end
+    	
+    local factor = LevelVars[dataName].Smooth and LevelFuncs.Engine.Node.Smoothstep(LevelVars[dataName].Progress) or LevelVars[dataName].Progress
+    local newValue1 = LevelFuncs.Engine.Node.Lerp(LevelVars[dataName].OldValue, LevelVars[dataName].NewValue, factor)
+    local object2 = TEN.Objects.GetMoveableByName(LevelVars[dataName].ObjectName)
+    local object2pos = object2:GetPosition()
+    local center2 = TEN.Objects.GetMoveableByName(LevelVars[dataName].CentreName)
+    local center2pos = center2:GetPosition()
+	local center2rot = center2:GetRotation()
```
Comment (Stranger1992) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua:66]: remove tabulation

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    local radius2 = LevelVars[dataName].Radius
+    local angle = newValue1 * math.pi / 180
+    local x = center2pos.x
+    local y = center2pos.y
+    local z = center2pos.z
+	local normalizedNewValue1 = NormalizeAngle(newValue1)
```
Comment (Stranger1992) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua:72]: remove tabulation

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    local normalizedEndAngle = NormalizeAngle(LevelVars[dataName].EndAngle)
+	
+
+    if (LevelVars[dataName].DataType == 0) then
+       	if (center2rot.y == 0) then
+
+			local ptx, pty = x + radius2 * math.sin(angle), y - radius2 * math.cos(angle)
```
Comment (Stranger1992) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua:80]: remove tabulation on line 78, 79, 80

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+			object2:SetPosition(Vec3(ptx, pty, object2pos.z))
+			if not IsSoundPlaying(LevelVars[dataName].Sfx) then
+            TEN.Sound.PlaySound(LevelVars[dataName].Sfx, object2pos)
```
Comment (Stranger1992) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua:84]: Add new line here

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        end
+        if LevelVars[dataName].CRotate then    
+            center2:SetRotation(Rotation(0, newValue1, 0))
+        end
+        if LevelVars[dataName].Rotate then    
+            object2:SetRotation(Rotation(0, newValue1, 0))
+        end
+    end
+
+    if LevelVars[dataName].StopAtEnd and  math.abs(normalizedNewValue1 - normalizedEndAngle) < 5e-1 then
```
Comment (Stranger1992) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: What is that `5e-1` ? Again, declare these magic numbers.

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+        LevelVars[dataName] = nil
+    end
+end
+
+--- Delete timed transform data
+LevelFuncs.Engine.Node.DeleteTimedData = function(objectName, endangle)
+    local dataName = objectName .. "_revolve_data"
+
+	if LevelVars[dataName] and LevelVars[dataName].Timer and LevelVars[dataName].Timer:IsActive() then
+   	LevelVars[dataName].EndAngle = endangle
```
Comment (Stranger1992) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: `endangle` should be `endAngle`

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+    LevelVars[dataName].Timer      = Timer.Create(dataName, 1 / 30, true, false, LevelFuncs.Engine.Node.TransformTimedData, dataName)
+    LevelVars[dataName].StartAngle = startangle
+    LevelVars[dataName].EndAngle   = endangle
+    LevelVars[dataName].CCW        = isCCW
+    LevelVars[dataName].Loop       = isLoop
+    LevelVars[dataName].Smooth     = isSmooth
+    LevelVars[dataName].Rotate     = isRotate
+    LevelVars[dataName].CRotate    = isCRotate
+    LevelVars[dataName].Sfx        = SFX
+    LevelVars[dataName].OldValue   = startangle
```
Comment (Stranger1992) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: `startAngle`

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+    LevelVars[dataName].Progress   = 0
+    LevelVars[dataName].Interval   = 1 / (time * 30)
+    LevelVars[dataName].DataType   = dataType
+    LevelVars[dataName].Radius     = radius
+    LevelVars[dataName].ObjectName = objectName
+    LevelVars[dataName].CentreName = objectcentre
+    LevelVars[dataName].Name       = dataName
+    LevelVars[dataName].Timer      = Timer.Create(dataName, 1 / 30, true, false, LevelFuncs.Engine.Node.TransformTimedData, dataName)
+    LevelVars[dataName].StartAngle = startangle
+    LevelVars[dataName].EndAngle   = endangle
```
Comment (Stranger1992) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: endAngle

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+		LevelVars[dataName] = nil
+	end
+
+	LevelVars[dataName] = {}
+    LevelVars[dataName].Progress   = 0
+    LevelVars[dataName].Interval   = 1 / (time * 30)
+    LevelVars[dataName].DataType   = dataType
+    LevelVars[dataName].Radius     = radius
+    LevelVars[dataName].ObjectName = objectName
+    LevelVars[dataName].CentreName = objectcentre
```
Comment (Stranger1992) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: objectCentre

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+    LevelVars[dataName].Name       = dataName
+    LevelVars[dataName].Timer      = Timer.Create(dataName, 1 / 30, true, false, LevelFuncs.Engine.Node.TransformTimedData, dataName)
+    LevelVars[dataName].StartAngle = startangle
+    LevelVars[dataName].EndAngle   = endangle
+    LevelVars[dataName].CCW        = isCCW
```
Comment (Stranger1992) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: Its not obvious what CCW is. Personally rename this symbol to `counterClockwise`

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+    LevelVars[dataName].Loop       = isLoop
+    LevelVars[dataName].Smooth     = isSmooth
+    LevelVars[dataName].Rotate     = isRotate
+    LevelVars[dataName].CRotate    = isCRotate
```
Comment (Stranger1992) [inline TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: Clarification needed on what CRotate is? Clockwise rotation? Clown rotation? Expand this symbol.

**Review by Stranger1992 — CHANGES_REQUESTED** [review]

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua (hunk 1 above)
Comment (TrainWrack) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua:75]: Done

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua (hunk 2 above)
Comment (TrainWrack) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: Done

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua (hunk 3 above)
Comment (TrainWrack) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: Done

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua (hunk 4 above)
Comment (TrainWrack) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua:66]: Done

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua (hunk 5 above)
Comment (TrainWrack) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua:72]: Done

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua (hunk 6 above)
Comment (TrainWrack) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua:80]: Done.

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua (hunk 7 above)
Comment (TrainWrack) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua:84]: Done

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua (hunk 8 above)
Comment (TrainWrack) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: This is a tolerance test. Declared.

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua (hunk 14 above)
Comment (TrainWrack) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: The explanation for the variable is in the node. Also added comments at variable declaration.

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua (hunk 13 above)
Comment (TrainWrack) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: The explanation for the variable is in the node. Also added comments at variable declaration.

**Comment by TrainWrack** [comment]
Please set the tab to 4 in the viewer. I am not really sure how to fix the tabs. The tabs look okay in the Notepad++.

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua (hunk 10 above)
Comment (TrainWrack) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: Done

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua (hunk 11 above)
Comment (TrainWrack) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: Done

File: TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua (hunk 12 above)
Comment (TrainWrack) [reply TombLib/TombLib/Catalogs/TEN Node Catalogs/Revolution.lua]: Done

**Review by Stranger1992 — APPROVED** [review]

**Pull request merged**